File: backend/routes/jobs.js
const express = require('express')
const router = express.Router()
const admin = require('../firebase-admin')

async function verifyToken(req, res, next) {
  const authHeader = req.headers.authorization
  if (!authHeader?.startsWith('Bearer ')) return res.status(401).json({ error: 'Unauthorized' })
  const token = authHeader.split('Bearer ')[1]
  try {
    const decoded = await admin.auth().verifyIdToken(token)
    req.uid = decoded.uid
    req.email = decoded.email
    next()
  } catch {
    res.status(401).json({ error: 'Invalid token' })
  }
}

// POST /api/jobs — create a new job listing
router.post('/', verifyToken, async (req, res) => {
  try {
    const db = admin.firestore()
    const job = {
      ...req.body,
      recruiterUid: req.uid,
      recruiterEmail: req.email,
      status: req.body.status || 'active',
      createdAt: admin.firestore.FieldValue.serverTimestamp(),
      updatedAt: admin.firestore.FieldValue.serverTimestamp(),
    }
    const ref = await db.collection('jobs').add(job)
    res.json({ success: true, jobId: ref.id })
  } catch (err) {
    console.error('Error creating job:', err)
    res.status(500).json({ error: 'Failed to create job' })
  }
})

// GET /api/jobs — fetch all non-draft jobs (for job feed)
router.get('/', async (req, res) => {
  try {
    const db = admin.firestore()
    const snap = await db.collection('jobs').where('status', '!=', 'draft').get()
    const jobs = snap.docs.map(doc => ({ id: doc.id, ...doc.data(), createdAt: doc.data().createdAt?.toDate?.()?.toISOString() || null }))
    jobs.sort((a, b) => (b.createdAt || '') > (a.createdAt || '') ? 1 : -1)
    res.json({ jobs })
  } catch (err) {
    console.error('Error fetching jobs:', err)
    res.status(500).json({ error: 'Failed to fetch jobs' })
  }
})

// GET /api/jobs/mine — fetch jobs posted by this recruiter
router.get('/mine', verifyToken, async (req, res) => {
  try {
    const db = admin.firestore()
    const snap = await db.collection('jobs').where('recruiterUid', '==', req.uid).get()
    const jobs = snap.docs.map(doc => ({ id: doc.id, ...doc.data(), createdAt: doc.data().createdAt?.toDate?.()?.toISOString() || null }))
    jobs.sort((a, b) => (b.createdAt || '') > (a.createdAt || '') ? 1 : -1)
    res.json({ jobs })
  } catch (err) {
    console.error('Error fetching recruiter jobs:', err)
    res.status(500).json({ error: 'Failed to fetch jobs' })
  }
})

// PUT /api/jobs/:id — update existing job
router.put('/:id', verifyToken, async (req, res) => {
  try {
    const db = admin.firestore()
    const ref = db.collection('jobs').doc(req.params.id)
    const doc = await ref.get()
    if (!doc.exists) return res.status(404).json({ error: 'Job not found' })
    if (doc.data().recruiterUid !== req.uid) return res.status(403).json({ error: 'Forbidden' })
    const { id, recruiterUid, recruiterEmail, createdAt, ...updates } = req.body
    await ref.update({ ...updates, updatedAt: admin.firestore.FieldValue.serverTimestamp() })
    res.json({ success: true, jobId: req.params.id })
  } catch (err) {
    console.error('Error updating job:', err)
    res.status(500).json({ error: 'Failed to update job' })
  }
})

// GET /api/jobs/:id — fetch single job by ID
router.get('/:id', async (req, res) => {
  try {
    const db = admin.firestore()
    const doc = await db.collection('jobs').doc(req.params.id).get()
    if (!doc.exists) return res.status(404).json({ error: 'Job not found' })
    const job = { id: doc.id, ...doc.data(), createdAt: doc.data().createdAt?.toDate?.()?.toISOString() || null }
    res.json({ job })
  } catch (err) {
    console.error('Error fetching job:', err)
    res.status(500).json({ error: 'Failed to fetch job' })
  }
})

// POST /api/jobs/:id/apply — employee applies for a job
router.post('/:id/apply', verifyToken, async (req, res) => {
  try {
    const db = admin.firestore()
    const jobRef = db.collection('jobs').doc(req.params.id)
    const jobDoc = await jobRef.get()
    if (!jobDoc.exists) return res.status(404).json({ error: 'Job not found' })
    const job = jobDoc.data()

    // Prevent duplicate applications
    const existing = await db.collection('applications')
      .where('jobId', '==', req.params.id)
      .where('applicantUid', '==', req.uid)
      .get()
    if (!existing.empty) return res.status(409).json({ error: 'Already applied' })

    const application = {
      jobId: req.params.id,
      jobTitle: job.title,
      hospital: job.hospital || job.companyName || '',
      recruiterUid: job.recruiterUid,
      recruiterEmail: job.recruiterEmail,
      applicantUid: req.uid,
      applicantEmail: req.email,
      applicantName: req.body.applicantName || '',
      coverNote: req.body.coverNote || '',
      status: 'New',
      appliedAt: admin.firestore.FieldValue.serverTimestamp(),
    }
    const ref = await db.collection('applications').add(application)
    res.json({ success: true, applicationId: ref.id })
  } catch (err) {
    console.error('Error applying for job:', err)
    res.status(500).json({ error: 'Failed to apply' })
  }
})

// PATCH /api/jobs/:id/status — pause/activate a job
router.patch('/:id/status', verifyToken, async (req, res) => {
  try {
    const db = admin.firestore()
    const ref = db.collection('jobs').doc(req.params.id)
    const doc = await ref.get()
    if (!doc.exists || doc.data().recruiterUid !== req.uid) return res.status(403).json({ error: 'Forbidden' })
    await ref.update({ status: req.body.status, updatedAt: admin.firestore.FieldValue.serverTimestamp() })
    res.json({ success: true })
  } catch (err) {
    res.status(500).json({ error: 'Failed to update job status' })
  }
})

// DELETE /api/jobs/:id — delete a job
router.delete('/:id', verifyToken, async (req, res) => {
  try {
    const db = admin.firestore()
    const ref = db.collection('jobs').doc(req.params.id)
    const doc = await ref.get()
    if (!doc.exists || doc.data().recruiterUid !== req.uid) return res.status(403).json({ error: 'Forbidden' })
    await ref.delete()
    res.json({ success: true })
  } catch (err) {
    res.status(500).json({ error: 'Failed to delete job' })
  }
})

module.exports = router

File: backend/src/api/v1/endpoints/jobs.py
"""src/api/v1/endpoints/jobs.py

Endpoints for job listing management and job applications.

Routes:
  POST   /api/jobs               — create a new job listing
  GET    /api/jobs               — list all non-draft jobs (public job feed)
  GET    /api/jobs/mine          — list jobs posted by this recruiter
  GET    /api/jobs/{id}          — get a single job by ID
  PUT    /api/jobs/{id}          — update a job listing (recruiter only)
  PATCH  /api/jobs/{id}/status   — pause or activate a job (recruiter only)
  DELETE /api/jobs/{id}          — delete a job listing (recruiter only)
  POST   /api/jobs/{id}/apply    — candidate applies for a job
"""

from fastapi import APIRouter, Depends, HTTPException, status
from google.cloud.firestore_v1 import SERVER_TIMESTAMP

from src.core.firebase import firebase_db
from src.core.security import get_current_user

router = APIRouter(prefix="/jobs", tags=["Jobs"])


def _serialize_job(doc) -> dict:
    d = doc.to_dict()
    created_at = d.get("createdAt")
    return {
        "id": doc.id,
        **d,
        "createdAt": created_at.isoformat() if hasattr(created_at, "isoformat") else None,
    }


@router.post("", summary="Create a new job listing")
async def create_job(body: dict, user: dict = Depends(get_current_user)):
    db = firebase_db()
    job = {
        **body,
        "recruiterUid": user["uid"],
        "recruiterEmail": user["email"],
        "status": body.get("status", "active"),
        "createdAt": SERVER_TIMESTAMP,
        "updatedAt": SERVER_TIMESTAMP,
    }
    ref = db.collection("jobs").add(job)
    return {"success": True, "jobId": ref[1].id}


@router.get("", summary="List all non-draft jobs (public job feed)")
async def list_jobs():
    db = firebase_db()
    snap = db.collection("jobs").where("status", "!=", "draft").stream()
    jobs = [_serialize_job(doc) for doc in snap]
    jobs.sort(key=lambda j: j.get("createdAt") or "", reverse=True)
    return {"jobs": jobs}


@router.get("/mine", summary="List jobs posted by this recruiter")
async def list_my_jobs(user: dict = Depends(get_current_user)):
    db = firebase_db()
    snap = db.collection("jobs").where("recruiterUid", "==", user["uid"]).stream()
    jobs = [_serialize_job(doc) for doc in snap]
    jobs.sort(key=lambda j: j.get("createdAt") or "", reverse=True)
    return {"jobs": jobs}


@router.get("/{job_id}", summary="Get a single job by ID")
async def get_job(job_id: str):
    db = firebase_db()
    doc = db.collection("jobs").document(job_id).get()
    if not doc.exists:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Job not found")
    return {"job": _serialize_job(doc)}


@router.put("/{job_id}", summary="Update a job listing")
async def update_job(job_id: str, body: dict, user: dict = Depends(get_current_user)):
    db = firebase_db()
    ref = db.collection("jobs").document(job_id)
    doc = ref.get()
    if not doc.exists:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Job not found")
    if doc.to_dict().get("recruiterUid") != user["uid"]:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Forbidden")

    for field in ("id", "recruiterUid", "recruiterEmail", "createdAt"):
        body.pop(field, None)

    ref.update({**body, "updatedAt": SERVER_TIMESTAMP})
    return {"success": True, "jobId": job_id}


@router.patch("/{job_id}/status", summary="Pause or activate a job listing")
async def update_job_status(job_id: str, body: dict, user: dict = Depends(get_current_user)):
    db = firebase_db()
    ref = db.collection("jobs").document(job_id)
    doc = ref.get()
    if not doc.exists or doc.to_dict().get("recruiterUid") != user["uid"]:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Forbidden")
    ref.update({"status": body.get("status"), "updatedAt": SERVER_TIMESTAMP})
    return {"success": True}


@router.delete("/{job_id}", summary="Delete a job listing")
async def delete_job(job_id: str, user: dict = Depends(get_current_user)):
    db = firebase_db()
    ref = db.collection("jobs").document(job_id)
    doc = ref.get()
    if not doc.exists or doc.to_dict().get("recruiterUid") != user["uid"]:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Forbidden")
    ref.delete()
    return {"success": True}


@router.post("/{job_id}/apply", summary="Apply for a job")
async def apply_for_job(job_id: str, body: dict, user: dict = Depends(get_current_user)):
    db = firebase_db()
    job_doc = db.collection("jobs").document(job_id).get()
    if not job_doc.exists:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Job not found")
    job = job_doc.to_dict()

    existing = (
        db.collection("applications")
        .where("jobId", "==", job_id)
        .where("applicantUid", "==", user["uid"])
        .get()
    )
    if existing:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail="Already applied")

    application = {
        "jobId": job_id,
        "jobTitle": job.get("title", ""),
        "hospital": job.get("hospital", job.get("companyName", "")),
        "recruiterUid": job.get("recruiterUid"),
        "recruiterEmail": job.get("recruiterEmail"),
        "applicantUid": user["uid"],
        "applicantEmail": user["email"],
        "applicantName": body.get("applicantName", ""),
        "coverNote": body.get("coverNote", ""),
        "status": "New",
        "appliedAt": SERVER_TIMESTAMP,
    }
    ref = db.collection("applications").add(application)
    return {"success": True, "applicationId": ref[1].id}
